Zend/zend_closures.c

```diff
@@ -705,6 +705,8 @@ void zend_register_closure_ce(void) /* {{{ */
 	zend_ce_closure = register_class_Closure();
 	zend_ce_closure->create_object = zend_closure_new;
 	zend_ce_closure->default_object_handlers = &closure_handlers;
+	/* FIXME: Potentially infer ZEND_ACC_MAY_BE_CYCLIC during construction of
+	 * closure? static closures not binding by references can't be cyclic. */
```

Comment:
I'm afraid my comment was incorrect.

```php
class Test {
    public \Closure $c;
}

$test = new Test();
$c = static function () use ($test) {}; // acyclic
$test->c = $c; // now it's cyclic
unset($test);
gc_collect_cycles();
unset($c); // leaks
```

This is only ok if `$test` itself is acyclic. As mentioned previously, this check would currently be unsound because the object may become cyclic at a later point in time by adding dynamic properties. However, once we're on PHP 9, it should be ok!

Reply:
Might be worth mentioning: `readonly` classes don't permit dynamic properties. So readonly classes could benefit from this optimization immediately without waiting for PHP 9.0.

Reply:
Actually, your test case doesn't show any dynamic property usage @iluuu1994, wouldn't that still be an issue in non-readonly cases even in PHP 9.0?

Reply:
> Might be worth mentioning: readonly classes don't permit dynamic properties. So readonly classes could benefit from this optimization immediately without waiting for PHP 9.0.

I was referring to:

> This is only ok if $test itself is acyclic. As mentioned previously, this check would currently be unsound

I.e. it's safe to assume `Closure` is acyclic only if it captures values that aren't themselves acyclic. If `Test` may point to cyclic values, as `Closure` itself, then it won't be marked as acyclic.

> Actually, your test case doesn't show any dynamic property usage @iluuu1994, wouldn't that still be an issue in non-readonly cases even in PHP 9.0?

Yes, but it's not a priority since it's a rare case and unlikely to make a big difference.